src/Framework/ReuseableStringBuilder.cs

```diff
@@ -0,0 +1,346 @@
+        /// Append a string.
+        /// </summary>
+        internal ReuseableStringBuilder Append(string value)
+        {
+            LazyPrepare();
+            _borrowedBuilder.Append(value);
+            return this;
+        }
+
+        /// <summary>
+        /// Append a substring.
+        /// </summary>
+        internal ReuseableStringBuilder Append(string value, int startIndex, int count)
+        {
+            LazyPrepare();
+            _borrowedBuilder.Append(value, startIndex, count);
+            return this;
+        }
+
+        public ReuseableStringBuilder AppendSeparated(char separator, ICollection<string> strings)
+        {
+            LazyPrepare();
+
+            var separatorsRemaining = strings.Count - 1;
+
+            foreach (var s in strings)
+            {
+                _borrowedBuilder.Append(s);
+
+                if (separatorsRemaining > 0)
+                {
+                    _borrowedBuilder.Append(separator);
+                }
+
+                separatorsRemaining--;
+            }
+
+            return this;
+        }
+
+        public ReuseableStringBuilder Clear()
+        {
+            LazyPrepare();
+            _borrowedBuilder.Clear();
+            return this;
+        }
+
+        /// <summary>
+        /// Remove a substring.
+        /// </summary>
+        internal ReuseableStringBuilder Remove(int startIndex, int length)
+        {
+            LazyPrepare();
+            _borrowedBuilder.Remove(startIndex, length);
+            return this;
+        }
+
+        /// <summary>
+        /// Grab a backing builder if necessary.
+        /// </summary>
+        private void LazyPrepare()
+        {
+            if (_borrowedBuilder == null)
+            {
+                FrameworkErrorUtilities.VerifyThrow(_capacity != -1, "Reusing after dispose");
+
+                _borrowedBuilder = ReuseableStringBuilderFactory.Get(_capacity);
+                _borrowedWithCapacity = _borrowedBuilder.Capacity;
+            }
+        }
+
+        /// <summary>
+        /// A utility class that mediates access to a shared string builder.
+        /// </summary>
+        /// <remarks>
+        /// If this shared builder is highly contended, this class could add
+        /// a second one and try both in turn.
+        /// </remarks>
+        private static class ReuseableStringBuilderFactory
+        {
+            /// <summary>
+            /// Made up limit beyond which we won't share the builder
+            /// because we could otherwise hold a huge builder indefinitely.
+            /// This was picked empirically to save at least 95% of allocated data size.
+            /// This constant has to exactly 2^n (power of 2) where n = 4 ... 32
+            /// </summary>
+            /// <remarks>
+            /// This constant might looks huge, but rather that lowering this constant,
+            ///   we shall focus on eliminating of code which requires to create such huge strings.
```

Comment:
```suggestion
            /// This constant might looks huge, but rather than lowering this constant,
            /// we shall focus on eliminating code which requires creating such huge strings.
```